cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/SparkPackageUtils.java

```diff
@@ -462,32 +465,70 @@ private static SparkFramework prepareSparkFramework(SparkCompat sparkCompat,
     String archiveName = "spark.archive-" + sparkVersion + "-" + VersionInfo.getVersion() + ".zip";
     Location frameworkDir = locationFactory.create("/framework/spark");
     Location frameworkLocation = frameworkDir.append(archiveName);
-
-    if (!frameworkLocation.exists()) {
-      File archive = new File(tempDir, archiveName);
-      try {
-        try (ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
-          zipOutput.setLevel(Deflater.NO_COMPRESSION);
-          for (File file : getLocalSparkLibrary(sparkCompat)) {
-            zipOutput.putNextEntry(new ZipEntry(file.getName()));
-            Files.copy(file.toPath(), zipOutput);
-            zipOutput.closeEntry();
-          }
+    File archive = new File(tempDir, archiveName);
+    try {
+      try (
+        ZipOutputStream zipOutput = new ZipOutputStream(new BufferedOutputStream(new FileOutputStream(archive)))) {
+        zipOutput.setLevel(Deflater.NO_COMPRESSION);
+        for (File file : getLocalSparkLibrary(sparkCompat)) {
+          zipOutput.putNextEntry(new ZipEntry(file.getName()));
+          Files.copy(file.toPath(), zipOutput);
+          zipOutput.closeEntry();
         }
-
-        // Upload spark archive to the framework location
-        frameworkDir.mkdirs("755");
-
-        try (OutputStream os = frameworkLocation.getOutputStream("644")) {
-          Files.copy(archive.toPath(), os);
+      } 
+      Location writeInProgressLoc = frameworkDir.append("write_in_progress");
+      frameworkDir.mkdirs("755");
+      //Create writing file
+      boolean writeCompleted = false;
+      while (!writeCompleted) {
+        if (writeInProgressLoc.createNew()) {
```

Comment:
It's cleaner to have the logic like this:

```java
while (!writeCompleted) {
  if (!uploadToLocation(archive, frameworkLocation, writeLockLocation)) {
    writeCompleted = waitForLocation(frameworkLocation);
  }
}
```

and move the `createNew` call inside the `uploadToLocation` method.